File: day4_secure_container/python/day4_secure_container_part_2.py
#!/bin/python

import sys

REPEAT_PAIRS = ["00", "11", "22", "33", "44", "55", "66", "77", "88", "99"]
DECREASE_PAIRS = set(["90", "91", "92", "93", "94", "95", "96", "97", "98",
    "80", "81", "82", "83", "84", "85", "86", "87",
    "70", "71", "72", "73", "74", "75", "76",
    "60", "61", "62", "63", "64", "65",
    "50", "51", "52", "53", "54",
    "40", "41", "42", "43",
    "30", "31", "32",
    "20", "21",
    "10",
    ])

def count_if_valid(x):
    x_string = str(x)
    char_pairs = [x_string[i : i + 2] for i in range(0, len(x_string) - 1)]
    if DECREASE_PAIRS.intersection(char_pairs):
        return 0
    else:
        return 1 if contains_isolated_pair(x_string) else 0

def contains_isolated_pair(s):
    for p in REPEAT_PAIRS:
        try:
            o = 0
            while True:
                o = s.index(p, o)
                if (o == 0 or s[o - 1] != p[0]) and (o + 2 >= len(s) or s[o + 2] != p[1]):
                    return True
                o += 1
        except:
            pass
    return False

def count_codes(min_inclusive, max_inclusive):
    ret = 0
    n = min_inclusive
    while n <= max_inclusive:
        ret += count_if_valid(n)
        n += 1
    return ret

min_inclusive = int(sys.argv[1])
max_inclusive = int(sys.argv[2])

print(count_codes(min_inclusive, max_inclusive))

File: day4_secure_container/rust/day4_secure_container_part_2.rs
use std::env;

fn main() {
    let args: Vec<String> = env::args().collect();
    let min_inclusive: i32 = args[1].parse::<i32>().unwrap();
    let max_inclusive: i32 = args[2].parse::<i32>().unwrap();
    println!("{}", count_codes(min_inclusive, max_inclusive));
}

fn count_codes(min_inclusive: i32, max_inclusive: i32) -> i32 {
    let mut ret = 0;
    let mut n = min_inclusive;
    while n <= max_inclusive {
        ret += count_if_valid(n);
        n += 1;
    }
    ret
}

fn count_if_valid(x: i32) -> i32 {
    let x_string: String = x.to_string();
    let mut x_chars = x_string.chars();
    match x_chars.next() {
        None => 0,
        Some(first_char) => {
            let mut preceding_char = first_char;
            let mut run_length = 1;
            let mut found_run2 = false;
            let mut non_decreasing = true;
            loop {
                match x_chars.next() {
                    None => break,
                    Some(c) => {
                        if c < preceding_char {
                            non_decreasing = false;
                            break;
                        } else if c == preceding_char {
                            run_length += 1;
                        } else {
                            found_run2 = found_run2 || run_length == 2;
                            run_length = 1;
                        }
                        preceding_char = c;
                    }
                }
            }
            found_run2 = found_run2 || run_length == 2;
            match non_decreasing && found_run2 {
                false => 0,
                true => 1,
            }
        },
    }
}
